openhcl/virt_mshv_vtl/src/processor/tdx/mod.rs

```diff
@@ -1679,53 +1717,55 @@ impl UhProcessor<'_, TdxBacked> {
                     // Emulate the access.
                     self.emulate(
                         dev,
-                        self.backing.interruption_information.valid(),
+                        self.backing.interruption_information[intercepted_vtl].valid(),
                         intercepted_vtl,
                     )
                     .await?;
                 }
 
-                &mut self.backing.exit_stats.ept_violation
+                &mut self.backing.exit_stats[intercepted_vtl].ept_violation
             }
             VmxExit::TPR_BELOW_THRESHOLD => {
                 // Loop around to reevaluate the APIC.
-                &mut self.backing.exit_stats.tpr_below_threshold
+                &mut self.backing.exit_stats[intercepted_vtl].tpr_below_threshold
             }
             VmxExit::INTERRUPT_WINDOW => {
                 // Loop around to reevaluate the APIC.
-                &mut self.backing.exit_stats.interrupt_window
+                &mut self.backing.exit_stats[intercepted_vtl].interrupt_window
             }
             VmxExit::NMI_WINDOW => {
                 // Loop around to reevaluate pending NMIs.
-                &mut self.backing.exit_stats.nmi_window
+                &mut self.backing.exit_stats[intercepted_vtl].nmi_window
             }
             VmxExit::HW_INTERRUPT => {
                 // Check if the interrupt was triggered by a hardware breakpoint.
                 let debug_regs = self
-                    .access_state(Vtl::Vtl0)
+                    .access_state(intercepted_vtl.into())
                     .debug_regs()
                     .expect("register query should not fail");
 
                 // The lowest four bits of DR6 indicate which of the
                 // four breakpoints triggered.
                 breakpoint_debug_exception = debug_regs.dr6.trailing_zeros() < 4;
-                &mut self.backing.exit_stats.hw_interrupt
+                &mut self.backing.exit_stats[intercepted_vtl].hw_interrupt
             }
-            VmxExit::SMI_INTR => &mut self.backing.exit_stats.smi_intr,
-            VmxExit::PAUSE_INSTRUCTION => &mut self.backing.exit_stats.pause,
+            VmxExit::SMI_INTR => &mut self.backing.exit_stats[intercepted_vtl].smi_intr,
+            VmxExit::PAUSE_INSTRUCTION => &mut self.backing.exit_stats[intercepted_vtl].pause,
             VmxExit::TDCALL => {
                 // If the proxy synic is local, then the host did not get this
                 // instruction, and we need to handle it.
                 if self.backing.untrusted_synic.is_some() {
+                    assert_eq!(intercepted_vtl, GuestVtl::Vtl0);
                     self.handle_tdvmcall(dev, intercepted_vtl);
                 } else if self.partition.hide_isolation {
```

Comment:
>hide_isolation [](http://example.com/codeflow?start=41&length=14)

should we support guest VSM when hiding isolation?

Reply:
That's a @jstarks question

Reply:
But I think yes? Since we support guest VSM when we're not isolated at all.